File: typescript/69.ts
function mySqrt(x: number): number {
    if (x === 0 || x === 1) return x;

    let left = 1;
    let right = Math.ceil(x / 2);
    while (left <= right) {
        const mid = Math.ceil((left + right) / 2);
        const pow = mid * mid;
        if (pow === x) {
            return mid;
        } else if (pow > x) {
            right = mid - 1;
        } else {
            left = mid + 1;
        }
    }

    return right;
};

test('69', () => {
    expect(mySqrt(4)).toBe(2);
    expect(mySqrt(8)).toBe(2);
});

File: python/69.py
import unittest
from math import ceil


class Solution:
    def mySqrt(self, x: int) -> int:
        if x == 0 or x == 1:
            return x

        left, right = 1, ceil(x / 2)
        while left <= right:
            mid = ceil((left + right) / 2)
            pow = mid * mid
            if pow == x:
                return mid
            elif pow > x:
                right = mid - 1
            else:
                left = mid + 1

        return right


class TestSolution(unittest.TestCase):
    def test(self):
        solution = Solution()
        self.assertEqual(solution.mySqrt(4), 2)
        self.assertEqual(solution.mySqrt(8), 2)
        self.assertEqual(solution.mySqrt(9), 3)
